sftp/sftp.go

```diff
@@ -455,9 +493,47 @@ func (o *Object) Remote() string {
 	return o.remote
 }
 
-// Hash returns "" since SFTP (in Go or OpenSSH) doesn't support remote calculation of hashes
 func (o *Object) Hash(r fs.HashType) (string, error) {
-	return "", fs.ErrHashUnsupported
+	session, err := o.fs.sshClient.NewSession()
+	if err != nil {
+		o.fs.cachedHashes = nil // Something has changed on the remote system
+		return "", fs.ErrHashUnsupported
+	}
+
+	err = fs.ErrHashUnsupported
+	var outputBytes []byte
+	escapedPath := shellEscape(o.path())
+	if r == fs.HashMD5 {
+		outputBytes, err = session.Output("md5sum " + escapedPath)
+	} else if r == fs.HashSHA1 {
+		outputBytes, err = session.Output("sha1sum " + escapedPath)
+	}
+
+	if err != nil {
+		o.fs.cachedHashes = nil // Something has changed on the remote system
+		_ = session.Close()
+		return "", fs.ErrHashUnsupported
+	}
+
+	_ = session.Close()
+	str := parseHash(outputBytes)
+	return str, nil
+}
+
+var shellEscapeRegex = regexp.MustCompile(`[^A-Za-z0-9_.,:/@\n-]`)
+
+// Escape a string s.t. it cannot cause unintended behavior
+// when sending it to a shell.
+func shellEscape(str string) string {
```

Comment:
Can you write a test for this please as it is important for security?

Put it in sftp_internal_test.go. Here is a [nice example](https://github.com/ncw/rclone/blob/master/http/http_internal_test.go#L241) of how to write that sort of test.